src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1255,54 +1289,63 @@
-                                    :msg "compiling accumulate with join filter node"}]
-             (compile-expr id
-                           (compile-join-filter (:join-filter-expressions beta-node)
-                                                (:join-filter-join-bindings beta-node)
-                                                (:env beta-node))))
-           (:result-binding beta-node)
-           children
-           join-bindings
-           (:new-bindings beta-node))
+          (let [join-filter-expr (compile-join-filter (:join-filter-expressions beta-node)
+                                                      (:join-filter-join-bindings beta-node)
+                                                      (:env beta-node))]
+            (with-meta
+              (eng/->AccumulateWithJoinFilterNode
+               id
+               ;; Create an accumulator structure for use when examining the node or the tokens
+               ;; it produces.
+               {:accumulator (:accumulator beta-node)
+                ;; Include the original filter expressions in the constraints for inspection tooling.
+                :from (update-in condition [:constraints]
+                                 into (-> beta-node :join-filter-expressions :constraints))}
+               compiled-accum
+               (binding [*compile-ctx* {:condition condition
+                                        :join-filter-expressions (:join-filter-expressions beta-node)
+                                        :env (:env beta-node)
+                                        :msg "compiling accumulate with join filter node"}]
+                 (compile-expr id
+                               join-filter-expr))
+               (:result-binding beta-node)
+               children
+               join-bindings
+               (:new-bindings beta-node))
+              {:accum-expr accum-expr
+               :join-filter-expr join-filter-expr}))
 
           ;; All unification is based on equality, so just use the simple accumulate node.
-          (eng/->AccumulateNode
-           id
-           ;; Create an accumulator structure for use when examining the node or the tokens
-           ;; it produces.
-           {:accumulator (:accumulator beta-node)
-            :from condition}
-           compiled-accum
-           (:result-binding beta-node)
-           children
-           join-bindings
-           (:new-bindings beta-node))))
+          (with-meta
+            (eng/->AccumulateNode
+             id
+             ;; Create an accumulator structure for use when examining the node or the tokens
+             ;; it produces.
+             {:accumulator (:accumulator beta-node)
+              :from condition}
+             compiled-accum
+             (:result-binding beta-node)
+             children
+             join-bindings
+             (:new-bindings beta-node))
+            {:accum-expr accum-expr})))
```

Comment:
I think this relates to https://github.com/rbrush/clara-rules/pull/219#issuecomment-243805826

I do think these should be made into first-class attributes of something.  I think the metadata is sort of a "hack".

I'll echo the statement I made there
>I actually went with this upon first pass because it was least invasive to the clara.rules.compiler and the clara.rules.engine.

The metadata was a minimal addition to get this off the ground.  To think about this though, if we did want a non-metadata approach, would putting it on the node really be the correct place?  Perhaps that would be useful for debugging (I think I would have had some value from that in the past).  However, if we targeted the beta graph for serialization instead, then this wouldn't necessarily be node attributes.  

So again, I just went with this as a "getting the implementation off the ground" approach.  If you have strong reservations against that though, we can explore our alternatives.